readthedocs/embed/v3/tests/test_access.py

```diff
@@ -0,0 +1,120 @@
+        self.version.privacy_level = PUBLIC
+        self.version.save()
+        self.url = (
+            reverse("embed_api_v3") + "?url=https://docs.readthedocs.io/en/latest/"
+        )
+        self.content = """
+        <html>
+            <div role=main>
+                Content
+            </div>
+        </html>
+        """
+
+    def get(self, *args, **kwargs):
+        """Wrapper around ``client.get`` to be overridden in the proxied api tests."""
+        return self.client.get(*args, **kwargs)
+
+    def _mock_open(self, content):
+        @contextmanager
+        def f(*args, **kwargs):
+            read_mock = mock.MagicMock()
+            read_mock.read.return_value = content
+            yield read_mock
+
+        return f
+
+    def _mock_storage(self, storage_mock):
+        storage_mock.open.side_effect = self._mock_open(self.content)
+
+    def test_get_content_public_version_anonymous_user(self, storage_mock):
+        self._mock_storage(storage_mock)
+        self.client.logout()
+        resp = self.get(self.url)
+        self.assertEqual(resp.status_code, 200)
+
+    def test_get_content_private_version_anonymous_user(self, storage_mock):
+        self._mock_storage(storage_mock)
+        self.version.privacy_level = PRIVATE
+        self.version.save()
+
+        self.client.logout()
+
+        resp = self.get(self.url)
+        self.assertEqual(resp.status_code, 403)
+
+    def test_get_content_public_version_logged_in_user(self, storage_mock):
+        self._mock_storage(storage_mock)
+        self.client.force_login(self.user)
+
+        resp = self.get(self.url)
+        self.assertEqual(resp.status_code, 200)
+
+    def test_get_content_private_version_logged_in_user(self, storage_mock):
+        self._mock_storage(storage_mock)
+        self.version.privacy_level = PRIVATE
+        self.version.save()
+
+        self.client.force_login(self.user)
+
+        resp = self.get(self.url)
+        self.assertEqual(resp.status_code, 200)
+
+    @mock.patch.object(EmbedAPIBase, "_download_page_content")
+    def test_get_content_allowed_external_page(
+        self, download_page_content, storage_mock
+    ):
+        download_page_content.return_value = self.content
+
+        resp = self.get(
+            reverse("embed_api_v3") + "?url=https://docs.python.org/en/latest/"
+        )
+        self.assertEqual(resp.status_code, 200)
+
+    def test_get_content_not_allowed_external_page(self, storage_mock):
+        resp = self.get(reverse("embed_api_v3") + "?url=https://example.com/en/latest/")
+        self.assertEqual(resp.status_code, 400)
+
+
+@pytest.mark.proxito
+class TestProxiedEmbedAPIV3Access(TestEmbedAPIV3Access):
+    def get(self, *args, **kwargs):
+        return self.client.get(*args, HTTP_HOST="docs.readthedocs.io", **kwargs)
+
+    def test_get_content_private_version_logged_in_user(self):
+        """Proxied API on .org doesn't log in users."""
```

Comment:
Something is missing here?

Reply:
I'm overriding this test to be empty, so it's skipped. I have updated the docstring to make it more explicit.